src/codegen/spirv.zig

```diff
@@ -1107,70 +1107,94 @@ const DeclGen = struct {
-                return ptr_id;
+                return result_ptr_id;
+            },
+            .decl_ptr => |decl| {
+                const result_ptr_ty = try zcu.declPtr(decl).declPtrType(zcu);
+                return self.constantDeclRef(result_ptr_ty, decl);
             },
-            .eu_payload => unreachable, // TODO
-            .opt_payload => unreachable, // TODO
-            .comptime_field, .comptime_alloc => unreachable,
-            .elem => |elem_ptr| {
-                const parent_ptr_ty = Type.fromInterned(mod.intern_pool.typeOf(elem_ptr.base));
-                const parent_ptr_id = try self.constantPtr(parent_ptr_ty, Value.fromInterned(elem_ptr.base));
-                const index_id = try self.constInt(Type.usize, elem_ptr.index, .direct);
-
-                const elem_ptr_id = try self.ptrElemPtr(parent_ptr_ty, parent_ptr_id, index_id);
-
-                // TODO: Can we consolidate this in ptrElemPtr?
-                const elem_ty = parent_ptr_ty.elemType2(mod); // use elemType() so that we get T for *[N]T.
-                const elem_ptr_ty_id = try self.ptrType(elem_ty, self.spvStorageClass(parent_ptr_ty.ptrAddressSpace(mod)));
-
-                // TODO: Can we remove this ID comparison?
-                if (elem_ptr_ty_id == result_ty_id) {
-                    return elem_ptr_id;
+            .anon_decl_ptr => |ad| {
+                const result_ptr_ty = Type.fromInterned(ad.orig_ty);
+                return self.constantAnonDeclRef(result_ptr_ty, ad);
+            },
+            .eu_payload_ptr => @panic("TODO"),
+            .opt_payload_ptr => @panic("TODO"),
+            .field_ptr => |field| {
+                const parent_ptr_id = try self.derivePtr(field.parent.*);
+                const parent_ptr_ty = try field.parent.ptrType(zcu);
+                return self.structFieldPtr(field.result_ptr_ty, parent_ptr_ty, parent_ptr_id, field.field_idx);
+            },
+            .elem_ptr => |elem| {
+                const parent_ptr_id = try self.derivePtr(elem.parent.*);
+                const parent_ptr_ty = try elem.parent.ptrType(zcu);
+                const index_id = try self.constInt(Type.usize, elem.elem_idx, .direct);
+                return self.ptrElemPtr(parent_ptr_ty, parent_ptr_id, index_id);
+            },
+            .offset_and_cast => |oac| {
+                const parent_ptr_id = try self.derivePtr(oac.parent.*);
+                const parent_ptr_ty = try oac.parent.ptrType(zcu);
+                disallow: {
+                    if (oac.byte_offset != 0) break :disallow;
+                    // Allow changing the pointer type child only to restructure arrays.
+                    // e.g. [3][2]T to T is fine, as is [2]T -> [2][1]T.
+                    const src_base_ty = parent_ptr_ty.arrayBase(zcu)[0];
+                    const dest_base_ty = oac.new_ptr_ty.arrayBase(zcu)[0];
+                    if (src_base_ty.toIntern() != dest_base_ty.toIntern()) break :disallow;
```

Comment:
```suggestion
                    if (target.os.tag == .vulkan and src_base_ty.toIntern() != dest_base_ty.toIntern()) break :disallow;
```
I'm not sure in how far this is allowed for Vulkan, but for OpenCL its fine. I think for Vulkan we'll need to figure out some extra derivation steps here. For example, `[3][2]T -> T` can be resolved by `&value[0][0]`. Going the other way around is a little more tricky, but perhaps it can be solved by making the representation of `*[1]T` be the same as `*T` (and just not use an array here). Anyway, that is for later. For now its fine like this.